mono/metadata/class-init.c

```diff
@@ -3151,60 +3151,58 @@ mono_class_setup_vtable_general (MonoClass *klass, MonoMethod **overrides, int o
 
-			// If there is an explicit implementation, just use it right away,
-			// otherwise look for a matching method
-			if (override_im == NULL) {
-				int cm_index;
-				MonoMethod *cm;
-
-				// First look for a suitable method among the class methods
-				for (l = virt_methods; l; l = l->next) {
-					cm = (MonoMethod *)l->data;
-					TRACE_INTERFACE_VTABLE (printf ("    For slot %d ('%s'.'%s':'%s'), trying method '%s'.'%s':'%s'... [EXPLICIT IMPLEMENTATION = %d][SLOT IS NULL = %d]", im_slot, ic->name_space, ic->name, im->name, cm->klass->name_space, cm->klass->name, cm->name, interface_is_explicitly_implemented_by_class, (vtable [im_slot] == NULL)));
-					if (check_interface_method_override (klass, im, cm, TRUE, interface_is_explicitly_implemented_by_class, (vtable [im_slot] == NULL))) {
-						TRACE_INTERFACE_VTABLE (printf ("[check ok]: ASSIGNING"));
+			int cm_index;
+			MonoMethod *cm;
+
+			// First look for a suitable method among the class methods
+			for (l = virt_methods; l; l = l->next) {
+				cm = (MonoMethod *)l->data;
+				TRACE_INTERFACE_VTABLE (printf ("    For slot %d ('%s'.'%s':'%s'), trying method '%s'.'%s':'%s'... [EXPLICIT IMPLEMENTATION = %d][SLOT IS NULL = %d]", im_slot, ic->name_space, ic->name, im->name, cm->klass->name_space, cm->klass->name, cm->name, interface_is_explicitly_implemented_by_class, (vtable [im_slot] == NULL)));
+				if (check_interface_method_override (klass, im, cm, TRUE, interface_is_explicitly_implemented_by_class, (vtable [im_slot] == NULL))) {
+					TRACE_INTERFACE_VTABLE (printf ("[check ok]: ASSIGNING"));
+					vtable [im_slot] = cm;
+					/* Why do we need this? */
+					if (cm->slot < 0) {
+						cm->slot = im_slot;
+					}
+					if (conflict_map)
+						g_hash_table_remove(conflict_map, im);
+				}
+				TRACE_INTERFACE_VTABLE (printf ("\n"));
+				if (mono_class_has_failure (klass))  /*Might be set by check_interface_method_override*/
+					goto fail;
+			}
+			
+			// If the slot is still empty, look in all the inherited virtual methods...
+			if ((vtable [im_slot] == NULL) && klass->parent != NULL) {
+				MonoClass *parent = klass->parent;
+				// Reverse order, so that last added methods are preferred
+				for (cm_index = parent->vtable_size - 1; cm_index >= 0; cm_index--) {
+					MonoMethod *cm = parent->vtable [cm_index];
+					
+					TRACE_INTERFACE_VTABLE ((cm != NULL) && printf ("    For slot %d ('%s'.'%s':'%s'), trying (ancestor) method '%s'.'%s':'%s'... ", im_slot, ic->name_space, ic->name, im->name, cm->klass->name_space, cm->klass->name, cm->name));
+					if ((cm != NULL) && check_interface_method_override (klass, im, cm, FALSE, FALSE, TRUE)) {
+						TRACE_INTERFACE_VTABLE (printf ("[everything ok]: ASSIGNING"));
 						vtable [im_slot] = cm;
 						/* Why do we need this? */
 						if (cm->slot < 0) {
 							cm->slot = im_slot;
 						}
+						if (conflict_map)
+							g_hash_table_remove(conflict_map, im);
```

Comment:
I'm not sure about this one.

```csharp

interface I1 { void M1 () { Console.WriteLine ("M1"); }
interface I2 : I1 { void I1.M1 () { Console.WriteLine ("M2"); }

class C1 : I1 { public void M1 ()  { Console.WriteLine ("C1"); } // this one is okay; call to M1 will print C1

class C2 : C1 { } // this should be ok.  call to M1 will print C1

class C3 : C1, I2 { } // is this okay?   Will it print C1 or I2?

Reply:
It wil print C1 calling M1 from C2 and from C3. Using mono or coreclr.